web-common/src/components/button-toggle-group/ButtonToggleGroup.svelte

```diff
@@ -0,0 +1,126 @@
+<script lang="ts" context="module">
+  export type SubButtonKey = number | string;
+
+  export type ButtonGroupContext = {
+    registerSubButton?: (key: SubButtonKey) => void;
+    toggleSubButton?: (key: SubButtonKey) => void;
+    selectedKey?: Writable<SubButtonKey>;
+    firstKey?: Writable<SubButtonKey>;
+    lastKey?: Writable<SubButtonKey>;
+    disabledKeys?: SubButtonKey[];
+  };
+  export const buttonGroupContext: ButtonGroupContext = {};
+</script>
+
+<script lang="ts">
+  import { setContext, onDestroy } from "svelte";
+  import { writable, get, Writable } from "svelte/store";
+  import { createEventDispatcher } from "svelte";
+
+  // If selectionRequired is true, then a sub button must be selected at all times.
+  // In this case the button group behaves like a radio button.
+  // If selectionRequired is false, It is possible to have no sub button selected.
+  // If additionally, defaultKey is undefined, then no sub button is selected by default.
+  export let selectionRequired = false;
+
+  export let defaultKey: number | string = undefined;
+  export let disabledKeys: (number | string)[] = [];
+
+  const dispatch = createEventDispatcher();
+
+  const subButtons = [];
+  const selectedKey = writable(null);
+
+  // if a default key is provided, then select it if it is not disabled
+  // this Must be run reactively in case the default key changes or
+  //  the set of disabled keys changes
+  $: if (defaultKey !== undefined && !disabledKeys.includes(defaultKey)) {
+    selectedKey.update(() => defaultKey);
+  } else {
+    selectedKey.update(() => null);
+  }
```

Comment:
In most implementations that I've seen of generic UI components, a "default value" (defaultKey) is treated like an "initialValue" and is _not reacted_ to when it changes. I.e., it only matters what the first default value is, it would not overwrite the existing value when it changes.

At a minimum here, wouldn't you want to check that selectedKey isn't null before trying to apply a default? Isn't the default akin to a "fallback" key?

Reply:
If the user doesn't provide a default key, why set the selected key to null?

Reply:
Does this component support a "controlled" approach, or just "uncontrolled"? If I want the parent component using the ButtonGroup to control the state, how do I do that? It almost feels like "defaultKey" is the way to do that, in which case its not properly named. See this example from MUI https://mui.com/material-ui/react-toggle-button/#ToggleButtons.tsx

Reply:
Hmm yeah this is pretty muddy -- thanks for pushing me to look more deeply into it. I guess it's probably more flexible as a controlled component. I'll rework it along those lines.